### PR TITLE
Crowdplatform2

### PR DESCRIPTION
**Pull request by marcelhollerbach**
crowdplatform basic shit

### CODE REVIEW

**Comment by marcelhollerbach** [comment]
no review! problems detected!

from leander:
- make sure no duplicated tasks for experiment+platform
- getworkerid works now different

**Comment by marcelhollerbach** [comment]
time for review.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    /**
+     * Publish the given experiment on the platform.
+     * The method will update the database with the new public task
+     *
+     * @param name The name of the platform
+     * @param experiment The experiment to publish
+     * @return None if the platform does not exist
+     */
+    public Optional<CompletableFuture<Boolean>> publishTask(String name, Experiment experiment) throws TaskOperationException {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: this signature is not very elegant

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        if (tasksOps.searchTask(name,experiment.getId()).isPresent())
+            throw new TaskOperationException("Experiment is already published!");
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i already do that in the database, and my guarantees are better. But because i was smart, i never told anyone about it

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                        TaskRecord record = new TaskRecord();
+                        record.setExperiment(experiment.getId());
+                        record.setPlatformData(s);
+                        record.setStatus(TaskStatus.running);
+                        record.setCrowdPlatform(name);
+                        tasksOps.createTask(record);
+                    }
+                    return true;
+                }));
+    }
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i would suggest:

``` java
return getPlatform(name)
                .map(platform -> platform.publishTask(experiment))
                .orElse(CompletableFuture.supplyAsync(() -> {throw new IllegalArgumentException("platform not found");}))
                .thenApply(platformData -> {
                    TaskRecord record = new TaskRecord();
                    record.setExperiment(experiment.getId());
                    record.setPlatformData(platformData);
                    record.setStatus(TaskStatus.running);
                    record.setCrowdPlatform(name);
                    tasksOps.createTask(record);
                    return true;
                });
```

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+    /**
+     * Unpublish a given experiment from the given platform
+     *
+     * @param name The name of the platform
+     * @param experiment The experiment to unpublish
+     * @return None if the platform was not found, false if the unpublish failed and true if everything went fine
+     */
+    public Optional<CompletableFuture<Boolean>> unpublishTask(String name, Experiment experiment) throws TaskOperationException {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: same as above

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    String getName();
+
+    /**
+     * Publish a passed experiment on the platform
+     * @param experiment the object which should be published
+     * @return This should return a unique string which is used to identify the published experiment later.
+     *         Will return None if the publishing failed
+     */
+    CompletableFuture<String> publishTask(Experiment experiment);
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java]: i would return an Future which completed exceptionally if the publishing failed, this makes the api easier to use

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        TaskRecord record;
+
+        record = tasksOps.searchTask(name, experiment.getId()).
+                orElseThrow(() -> new TaskOperationException("Experiment is not published"));
+
+        return getPlatform(name).map(platform1 ->
+            platform1.unpublishTask(record.getPlatformData()).handle((b, throwable) -> {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: if you only need the result and not the error (if there was one), i would suggest thenApply

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+        );
+    }
+
+    /**
+     * update the given experiment on the given platform
+     * @param name The name of the platform
+     * @param experiment The experiment to update
+     * @return None if the platform was not found, false if the update failed and true if everything went fine.
+     */
+    public Optional<CompletableFuture<Boolean>> updateTask(String name, Experiment experiment) throws TaskOperationException {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: see the comments above

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+     *
+     * @param name The name of the platform
+     * @param experiment The experiment to unpublish
+     * @return None if the platform was not found, false if the unpublish failed and true if everything went fine
+     */
+    public Optional<CompletableFuture<Boolean>> unpublishTask(String name, Experiment experiment) throws TaskOperationException {
+        TaskRecord record;
+
+        record = tasksOps.searchTask(name, experiment.getId()).
+                orElseThrow(() -> new TaskOperationException("Experiment is not published"));
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i would suggest multiple unpublishing results in no changes, this would simplify the api

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        return getWorker(name).flatMap(worker -> worker.identifyWorker(params));
+    }
+
+    /**
+     * Get a worker record from the database which is associated with the given parameters
+     * @param name Name of the platform
+     * @param params Params which were passed by the platform to the workerservice
+     * @return A WorkerRecord if one is found
+     */
+    public Optional<WorkerRecord> getWorker(String name, Map<String, String[]> params) {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: the hasEmail (or whatever you would like to name it) is missing

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        return identifyWorker(name, params).map(uid ->
+            workerOps.getAllWorkers().stream().
+                    filter(workerRecord1 -> workerRecord1.getIdentification().equals(uid)).findFirst().
+                    orElseGet(() -> {
+                        //create a new entry in the database
+                        WorkerRecord record = new WorkerRecord(-1, uid, name, null);
+                        record = workerOps.createWorker(record);
+                        return record;
+                    })
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: it more readable if every stream command has its own line (also the dots are typically in the front):

``` java
workerOps.getAllWorkers().stream()
                   .filter(workerRecord1 -> workerRecord1.getIdentification().equals(did))
                   .findFirst()
                   .orElseGet(() -> {
                        //create a new entry in the database
                        WorkerRecord record = new WorkerRecord(-1, uid, name, null);
                        record = workerOps.createWorker(record);
                        return record;
                    })
```

i would also create a new workerOperations command to get one Worker getWorker(platform, identification)

``` java
workerOps.getWorker(platform, identification)
                  .orElseGet(() -> {
                        //create a new entry in the database
                        WorkerRecord record = new WorkerRecord(-1, uid, name, null);
                        record = workerOps.createWorker(record);
                        return record;
                    })
```

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        );
+    }
+
+    public Optional<CompletableFuture<Boolean>> payWorker(String name, Worker worker, int amount) {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: signature

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    /**
+     * Will return the Worker interface which should be used to identify workers for the given platform
+     *
+     * @param name The name of the platform
+     * @return The interface used to identify a worker
+     */
+    public Optional<Worker> getWorker(String name) {
+        return getPlatform(name).map(platform -> platform.getWorker().orElse(fallbackWorker));
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i would do a new line after every operation, but thats taste ;)

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+    /**
+     * Will return the payment service which should be used for the given platform
+     * If there is no Platform with the given name None is returned.
+     *
+     * @param name The name of the platform to use
+     * @return The interface used for payment
+     */
+    public Optional<Payment> getPlatformPayment(String name) {
+        return getPlatform(name).map(platform -> platform.getPayment().orElse(fallbackPayment));
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: above

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+     * @param experiment The experiment with the new parameters
+     * @return The new id
+     */
+    CompletableFuture<String> updateTask(String id, Experiment experiment);
+
+    /**
+     * Returns if calibration questions can be asked on this platform
+     * @return true or false
+     */
+    Boolean isCalibsAllowd();
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java]: strange abbreviation

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Worker.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Created by marcel on 19.01.16.
+ */
+public interface Worker {
+    /**
+     * Parse a worker id out of the params
+     * @param param The parameters which were sent by a platform
+     * @return The id of the worker if one can be found
+     */
+    Optional<String> identifyWorker(Map<String, String[]> param);
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Worker.java]: i would throw an exception instead of using an optional. something went terribly wrong if the Platform is unable to parse it's own input.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 2 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: where do you check this ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 9 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: hasEmail is redundant for me, since i will get a email if one of the interfaces is not passed, so i dont care about that flag.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java (hunk 14 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java]: tell it the one who patched the "entwurf ;)"

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Worker.java (hunk 15 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Worker.java]: yo.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 10 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: as you like, not a big pain for me.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 6 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: ok

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java (hunk 14 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java]: i would still not use it in the actual code

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 2 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: doIfNotRunning

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 9 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: no its not, if the platform needs an email and hasEmail is false no worker must be created

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 2 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: well this would mean i need to create the task -> check if not failed -> publish on crowdplatform -> remove if failed -> insert data if not -> update which is like 100 times harder than doing that. maybe make ifNotRunnign public as boolean so i can use it ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 9 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: do you have some writen spec on what this api should do ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 2 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: ah ok, yeah i forgot that. it would first create the task and then publish it on the crowdPlatform (and if failed remove it) because of concurrency

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 9 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: getWorker:
used by the WorkerService to get the internal representation of the worker it communicates with. It first searches for the worker in the database and if not found may return a new one (see hasEmail).
@param String platform: the crowdworking platform the worker is working in
@param Map<String, String[]> data: passed platform-specific data from the worker, may be used to identify the worker by the platform
@param boolean hasEmail: if the crowdplatform needs an email, but the workerservice does not have one right now the method should only return a worker if it is already existing (and has an email). if not the parameter is always false and the platform should create a new one

Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: this sounds like a poor api, why only create a worker under this special cases ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 2 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: where is a concurrency problem ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 9 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i was thinking about it. I don't think its that bad, just look at java.util.map :P
we can split it, but the implementation into getExistingWorker and getWorker, but the implementation of the method above is pretty trivial (just look at the concept, just guard the create worker block with `(hasNativePayment || hasEmail)`)

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 2 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: if you publish a task twice then on the if clause would fail for both requests and therefore not throw an exception. I don't what happens then with the whole system. Therefore you don`t check mutable things like that if you change them afterwards

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 9 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: what i am wondering about is: A crowdplatform should identify a worker. The logic under which cases a user should be created or just read is not the job of the platform management. Thats a implementation case of a special rest call or some implementation in the workerservice. 
tl;dr this logic does not belong in this place here.

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i mean you can move the logic around, but i think i may belong here. i mean the platform-package is concerned with platform-stuff and this is platform stuff, independent of the db-call.  But we can move the db logic into the rest-call area and create a getPlatformData method that takes the platform and returns the workerRecord and another that returns the string. (i would really like to not be so dependent of the exact matching of the string, because thats more the limitations of the current database and we may get rid of it in the future, giving us far greater flexibility)

Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: the platform package is concerned with platform stuff, true. the only platform stuff here is the identification. The reason why i think this does not belong here: lets say the ws is changing, also the rest call. you would have to change the platform package, which is somehow wrong, nothing in the platform has changed, only the rest call and the ws.

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i mean my idea above about getWorker and createPlatformData (maybe better than get). They are not rest calls.

Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: sure, they are not direct rest calls, but the logic of the call is very depending on the way how the ws works. And has no basic usage like publishTask or something like that. The way how the method works is very depending on the ws logic. Thats why I would like to bind this logic to the rest call, will add a patch if this branch is landed.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java (hunk 14 above)
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java]: Misses an `e` in `Allowed`, too.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 9 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: getWorker and createPlatformData data is basic logic, you can't divide it up to a more primitive call. The rest call combines these primitives to whatever the deepest desires of the ws fulfils. Also its purely platform-dependent, i can`t do anything in the rest-package

Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: well for the platform the basic logic is to identify the worker. A call like identifyWorker. The logic in which cases a new record is created or not can be done in the rest call.

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: getWorker gets a worker if he is existing, createPlatformData would just generate what to save with the data. only queries, no inserts if thats so important to you. I really don`t want to rely on string equality in the api that much (therefore getWorker), it will limit future flexibility heavily.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java (hunk 14 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Platform.java]: thank you captain obvious

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 9 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: I added the code, and in the end in the two functions nothing else than db operations are done, which are not really something which needs to be done in the manager. Those db calls could have also been called in the rest handler.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+            throw new TaskOperationException("Task could not be created");
+
+        return getPlatform(name)
+                .map(platform1 -> platform1.publishTask(experiment))
+                .orElseThrow(() -> new IllegalArgumentException("Platform not found!"))
+                .handle((s1, throwable) -> {
+                    if (s1 != null) {
+                        result.setPlatformData(s1);
+                    } else {
+                        result.setStatus(TaskStatus.stopped);
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: you have to call method in the operations package. This will change nothing in the database and i don`t guarantee that .store will work

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+     * Unpublish a given experiment from the given platform
+     *
+     * @param name The name of the platform
+     * @param experiment The experiment to unpublish
+     * @return None if the platform was not found, false if the unpublish failed and true if everything went fine
+     */
+    public CompletableFuture<Boolean> unpublishTask(String name, Experiment experiment) throws TaskOperationException {
+        TaskRecord record;
+
+        record = tasksOps.searchTask(name, experiment.getId()).orElse(null);
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i would first set it to unpublished, so no other thread can unpublish it at the same time

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+        TaskRecord record;
+
+        record = tasksOps.searchTask(name, experiment.getId()).
+                orElseThrow(() -> new TaskOperationException("Experiment is not published"));
+
+        return getPlatform(name)
+                .map(platform -> platform.updateTask(record.getPlatformData(), experiment))
+                .orElseThrow(() -> new IllegalArgumentException("Platform not found"))
+                .thenApply(s -> {
+                    record.setPlatformData(s);
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:185]: same as above

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    /**
+     * Get a worker if he exists
+     * @param name Name of the platform
+     * @param params Params passed by the platform
+     * @return A worker if one is found
+     * @throws UnknownWorkerException if the platform does not identify a worker
+     */
+    public Optional<WorkerRecord> getWorker(String name, Map<String ,String[]> params) throws UnknownWorkerException {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i don`t get the signature, why not return empty if not found?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 16 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: wad ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 18 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:185]: updateTask does _not_ update stuff ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Worker.java [hunk 20]
```diff
@@ -0,0 +1,15 @@
+package edu.kit.ipd.crowdcontrol.objectservice.crowdworking;
+
+import java.util.Map;
+
+/**
+ * Created by marcel on 19.01.16.
+ */
+public interface Worker {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/Worker.java]: i don`t think the name is optimal, i would name the interface IdentifyWorker and also mark it with @FunctionalInterface.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 19 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: it returns none if not found, and the exception if the platfrom can not identify the user

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 16 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: this class only holds the results of the query. you can manipulate the data in TaskRecord, but that won't change the database.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 18 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:185]: no, i mean record stuff not changing the database

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 16 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: i recommend reading the next lines of code

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 19 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: that seems a bit clunky, why not return empty if no worker was found? and if you really want to use the exceptions, it should be the other way around (UnkownPlatformException and empty worker if worker not found)

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 16 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i have read it, but you whole method never changes anything in the database

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: i have read it, but your method publishes the task, but never updates its content. The platformData will be lost when the method returns

Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: i have absolutly no idea what you are talking about. if                     

```
record.setPlatformData(s);
return tasksOps.updateTask(record);
```

Do you want to tell me this is not updating the database ?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 19 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: empty is returned if not worker is found in the database, and a exception if the platform failes to identify the user (which means the whole request is faulty)

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 16 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: yeah, the record instances are the results of queries. They are mutable but changing their data does not change the data in the database. They only represent the results. There is a method .store, but i don't guarantee that it's working, so don't use it.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 19 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: ah ok, then i would rename the Exception

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 16 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: ok - you tell me updateTask does not update the databse and .store maybe but maybe not. So how do I use this magic bit called db ? :>

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: just create a method in the database package, like updateTask(TaskRecord record)

Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: WAD ? like the one which exist and is called 2 lines later ? WTF

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 19 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: no, just read the comment in which case the exception is thrown then there is no need to rename the exception

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 16 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: ok i am now officially stupid and unable to review code. yeah, this is the method :D I should probably take a break...

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 18 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:185]: ok, also my fault

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 16 above)
Comment (marcelhollerbach) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java:137]: go and get some air!!! :+1:

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 19 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: no, the name of the execution must describe its purpose. often these exceptions get propagated through the system so you have to rely on good names to understand where the exception is coming from and what it is doing.

**Comment by LeanderK** [comment]
i think it's good to go, but i don`t trust my decisions right now. I will go for a run and then sit down and review it seriously, so you hear from me in 45 minutes. I think the residual alcohol from yesterday is not helping either.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java (hunk 19 above)
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: I think `UnknownWorkerException` is fine, `WorkerIdentificationFailedException` might be an alternative name.

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: i think the new name is better

Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/crowdworking/PlatformManager.java]: What's the new / old name?
